gpflow/conditionals/util.py

```diff
@@ -51,11 +51,11 @@ def base_conditional(Kmn: tf.Tensor,
         0)  # [N]
     Kmn = tf.transpose(Kmn, perm)  # [..., M, N]
 
-    leading_dims = Kmn.shape[:-2]
+    leading_dims = tf.shape(Kmn)[:-2]
```

Comment:
Does the `[:-2]` still work in graph mode (wrapped in `@tf.function`)? :thinking:

Reply:
No I'd be happy with it I just vaguely remember some issues with slicing tf.shape() results in tf1 not working properly... so wanted to make sure it does work. (-> we need tests that wrap code in `tf.function()` and check it works the same as without...)

I can't access your mwe link - can you put it as a github gist or post in here?

Reply:
[here](https://gist.github.com/marcoadurno/97fc32c4387d2c77b97cbba9cab04b55)?

Reply:
Regarding the tests, similar to Example Usage [here](https://www.tensorflow.org/api_docs/python/tf/function#used_in_the_tutorials)?

Shall the tests be tied up with issue #1185 ?

Reply:
I think we should add tf.function with input_signature set to `tf.TensorSpec(shape=None)` in the tests.

This is distinct from #1185 where the intent is just to add `tf.function()` in the way one would usually use it (without passing `input_signature`, so not testing whether it would work with minibatching most of the time).